rectools/visuals/visual_app.py

```diff
@@ -0,0 +1,273 @@
+import typing as tp
+
+import ipywidgets as widgets
+import numpy as np
+import pandas as pd
+from IPython.display import display
+
+from rectools import Columns
+
+
+class AppDataStorage:
+    """
+    Storage and processing of data for `VisualApp` widgets. This class is not meant to be used
+    directly. Use `VisualApp` class instead
+    """
+
+    def __init__(
+        self,
+        recos: tp.Dict[tp.Hashable, pd.DataFrame],
+        item_data: pd.DataFrame,
+        requests_dict: tp.Dict[tp.Hashable, tp.Hashable],
+        is_u2i: bool = True,
+        interactions: tp.Optional[pd.DataFrame] = None,
+    ) -> None:
+        self.request_colname = Columns.User if is_u2i else Columns.TargetItem
+        self.is_u2i = is_u2i
+        self.requests_dict = requests_dict
+        self.model_names = list(recos.keys())
+        self.request_names = list(requests_dict.keys())
+
+        self._check_columns_present_in_recos(recos)
+        self.recos = recos
+
+        if Columns.Item not in item_data:
+            raise KeyError(f"Missed {Columns.Item} column in item_data")
+        self.item_data = item_data
+
+        if interactions is None and is_u2i:
+            raise ValueError("For u2i reco you must specify interactions")
+        if interactions is not None and not is_u2i:
+            raise ValueError("For i2i reco you shouldn't specify interactions")
+        if not is_u2i:
+            interactions = self._prepare_interactions_for_i2i()
+        self.interactions: pd.DataFrame = interactions
+
+        if not self.requests_dict:
+            raise ValueError("`requests_dict` is empty")
+        self.processed_interactions = self._process_interactions()
+        self.processed_recos = self._process_recos()
+
+    def _process_interactions(self) -> tp.Dict[tp.Hashable, pd.DataFrame]:
+        prepared_interactions = {}
+        for request_name, request_id in self.requests_dict.items():
+            prepared_interactions[request_name] = (
+                self.interactions[self.interactions[self.request_colname] == request_id]
+                .merge(self.item_data, how="left", on=Columns.Item)
+                .drop(columns=[self.request_colname])
+            )
+        return prepared_interactions
+
+    def _process_recos(self) -> tp.Dict[tp.Hashable, tp.Dict[tp.Hashable, pd.DataFrame]]:
+        prepared_recos = {}
+        for model_name, full_recos in self.recos.items():
+            model_recos = {}
+            for request_name, request_id in self.requests_dict.items():
```

Comment:
Don't you think it can be more effective to
1. Filter all request ids using `isin` from `full_recos`
2. merge
3. separate
?

Same for interactions

Reply:
Yes, it gives small boost for large dataframes. Fixed